MQTT over Websocket

jukkar opened the pull request:
Initial support for running MQTT over Websocket. The mqtt_publisher sample application in this PR, is able to connect to Websocket enabled mosquitto MQTT server.

Note that this PR contains commits from #19025 and all the comments about Websocket or HTTP APIs / samples should be written there.

jukkar committed:
net: mqtt: Add support for running MQTT over Websocket

Initial support for running MQTT over Websocket.

Fixes #19539

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
samples: net: mqtt_publisher: Add support for Websocket

Add a possibility to run MQTT over Websocket.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
samples: net: mqtt_publisher: Enable net shell

Network shell is useful to have in order to debug things so
enabling it for this sample application.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

rlubos commented on subsys/net/lib/mqtt/mqtt_transport_websocket.c:
```diff
@@ -0,0 +1,159 @@
+		flags |= MSG_DONTWAIT;
+	}
+
+	ret = recv(client->transport.websocket.sock, data, buflen, flags);
```
I think we should be consistent in terms of which API we use (`mqtt_client_websocket_write` calls `websocket_send_msg` while `mqtt_client_websocket_read` calls `recv`).

Personally, I'd stick to the `websocket_*` API, as from my point of view, it should make it easier to get this transport integrated with offloading. After moving the offloading underneath `zsock_*` (hopefully, I could look into this topic soon), both APIs should work, so I won't object if you decide to use socket API, but for now `websocket_*` is the only viable option for offloading.

jukkar replied:
Agreed, I will change the code to use websocket_* API.

Note that the websocket API uses socket interface so the offloading case should be automatically handled.

rlubos commented on subsys/net/lib/mqtt/mqtt_transport_websocket.c:
```diff
@@ -0,0 +1,159 @@
+#include "mqtt_os.h"
+#include "mqtt_transport.h"
+
+/**@brief Handles connect request for Websocket transport.
```
I have a feeling that we keep duplicating these docs needlessly. This functionality is documented in `mqtt_transport.h`, so it's probably not needed to repat it in specific transport implementations. Personally, I'd remove the docs from other transports as well.

jukkar replied:
Agreed, I will add a patch to this PR that does that.

rlubos reviewed (commented):
Looks pretty good, left a few remarks to consider.

dbkinder reviewed (approved):
doc changes LGTM.

jukkar commented on subsys/net/lib/mqtt/mqtt_transport.h:
```diff
@@ -95,6 +95,34 @@ int mqtt_transport_read(struct mqtt_client *client, u8_t *data, u32_t buflen,
 int mqtt_transport_disconnect(struct mqtt_client *client);
 
+/* Transport handler functions for TCP socket transport. */
+extern int mqtt_client_tcp_connect(struct mqtt_client *client);
```
Yep, I will remove that.

jukkar committed:
net: mqtt: Remove extra documentation from individual transports

The MQTT transport API functions are already documented in
mqtt_transport.h so need to duplicate them in individual
transport .c file.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
samples: net: mqtt_publisher: Use logging macros for output

Instead of printk(), use logging macros so that all the output
from IP stack and sample is nicely interleaved.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar commented:
Fixed the connectivity issues to mosquitto. The MQTT over Websocket works now in the latest version.

rlubos reviewed (approved):
LGTM

jukkar commented:
Lets merge #19025 before merging this one. I will keep this PR updated as needed.

jukkar commented:
Updated according to changes in #19025
